layers/+readers/djvu/README.org

```diff
@@ -0,0 +1,87 @@
+  - [[#features][Features:]]
+- [[#install][Install]]
+  - [[#prerequisites][Prerequisites]]
+  - [[#dotfile][Dotfile]]
+- [[#key-bindings][Key bindings]]
+
+* Description
+This layer adds support for reading djvu files with spacemacs.
+
+Because of some difficulties, the layer does not lazy load the =djvu.el= and
+=djvu3.el= packages.
+
+** Features:
+This layer provides a full featured djvu editor by implementing the [[https://github.com/dalanicolai/djvu2.el][djvu.el
+package]] along with the [[https://github.com/dalanicolai/djvu3][djvu3]] extension (a newer alternative to [[https://github.com/dalanicolai/djvu2.el][djvu2.el]]).
+
+- flexible annotation editing and rendering, covering most (but not all) of
+  [[https://linux.die.net/man/1/djvused][djvused]] annotation definitions
+- fast navigation with imenu
+- fast search with djvu-occur
+- remember last view with djvu-restore
+- dark mode
+
+* Install
+** Prerequisites
+  The =djvu= package requires the =djvused= command (from [[http://djvu.sourceforge.net/][djvulibre]]) to be
+  available in =exec-path=.
+
+  Additionally, for =dark-mode= availability of the command =pnminvert= (from
+  [[http://netpbm.sourceforge.net/][netpbm]]) is required.
+  
+** Dotfile
+    After installing the dependencies, add this to your ~/.spacemacs.
+
+#+BEGIN_SRC emacs-lisp
+(setq-default dotspacemacs-configuration-layers '(djvu))
+#+END_SRC
+
+After that, syncronize your configuration with SPC f e R.
+
+* Key bindings
+
+*djvu-read-mode*
+| Key Binding | Description                                                                      |
+|-------------+----------------------------------------------------------------------------------|
+| ~i~         | djvu-image-toggle                                                                |
+| ~j~         | scroll up                                                                        |
+| ~k~         | scroll down                                                                      |
+| ~J~         | next page                                                                        |
+| ~K~         | previous page                                                                    |
+| ~g~         | goto page                                                                        |
+|             |                                                                                  |
+| ~d~         | djvu-toggle-invert (dark-mode)                                                   |
+| ~c~         | djvu-toggle-semi-continuous-scrolling (works only when image larger than window) |
+| ~+/-~       | zoom-in/out                                                                      |
+| ~, s~       | djvu-occur                                                                       |
+| ~/~         | djvu fast search                                                                 |
+| ~n~         | djvu search continue                                                             |
```

Comment:
https://github.com/dalanicolai/djvu2.el/blob/c89f513e589e06bcf239db0f502ed9bac34753d1/djvu.el#L2198-L2210

this underlying implementation can be improved by keeping a counter `i`, and when the loop end, `(djvu-next-page i)`.

Reply:
I am not using the file djvu2.el in this PR. This PR contains djvu.el (from ELPA) with quite some extensions in djvu3.el and some improved functions in funcs.el. I have tried to get the improvements into djvu.el but there was neither a response to my message to the [bug-gnu-emacs](https://lists.gnu.org/archive/html/bug-gnu-emacs/2021-05/msg01755.html) nor to my message to the [emacs-devel](https://lists.gnu.org/archive/html/emacs-devel/2021-05/msg00944.html) mailing list.

Anyway, I haven't bound the function you are pointing to (the original function from djvu.el), because I have added a faster simple search version in `funcs.el`. But as I mentioned, I did not bother to improve it further because users really should use the `djvu-occur` function, if they prefer even better search functionality.

Reply:
well, except a few limited cases, a symbol in package `foo` should be prefixed with `foo-`.
so you need to rename the symbols in `djvu3`, or otherwise it's really confusing.

Reply:
Ah okay, well I had named them djvu because they were destined to get merged into djvu.el. Also, I am overriding functions of djvu.el, so for that I should really use the same name of the function in djvu.el (or create an alias), because these functions are used in djvu.el itself. Anyway, if the PR gets merged, then when users jump to the definition of the functions, they will end up in djvu3.el, so I think it is not too confusing (although it is a little unfortunately). Anyway, `djvu2.el` was not involved in this PR at all.